NeurodataWithoutBorders/pynwb
# add custom html field generation method for TimeSeries

## Motivation

Fix #1830. 

Adds a method to TimeSeries that checks for linked timestamps or linked data to avoid recursion errors when generating an html representation in Jupyter notebooks.

See also related hdmf issue (https://github.com/hdmf-dev/hdmf/issues/1010) and PR (https://github.com/hdmf-dev/hdmf/pull/1038)

## How to test the behavior?
See below for an example from the related hdmf issue in which timestamps are linked across multiple spatial series.

```
from pynwb import NWBHDF5IO
from pynwb.testing.mock.file import mock_NWBFile
from pynwb.testing.mock.behavior import mock_SpatialSeries, mock_Position
import numpy as np

nwbfile = mock_NWBFile()

test_timestamps = np.zeros((1000,))
test_data = np.random.rand(*(1000, 3))
spatial_series1 = mock_SpatialSeries(name="test1", rate=None, data=test_data, timestamps=test_timestamps)
spatial_series2 = mock_SpatialSeries(name="test2", rate=None, data=test_data, timestamps=spatial_series1)
spatial_series3 = mock_SpatialSeries(name="test3", rate=None, data=test_data, timestamps=spatial_series1)
spatial_series4 = mock_SpatialSeries(name="test4", rate=None, data=test_data, timestamps=spatial_series1)
position = mock_Position(spatial_series=[spatial_series1, spatial_series2, spatial_series3, spatial_series4])

nwbfile.create_processing_module("behavior", description="contains processed behavior data")
nwbfile.processing["behavior"].add(position)

with NWBHDF5IO("test.nwb", "w") as io:
    io.write(nwbfile)

read_io = NWBHDF5IO("test.nwb", "r")
nwbfile_in = read_io.read()

nwbfile_in
```

## Checklist
